packages/wallet/api/src/marshal-contexts.js

```diff
@@ -0,0 +1,257 @@
+        const id = myData[kind].byVal.get(val);
+        return `${kind}:${id}`;
+      }
+    }
+    unknownNonce += 1;
+    const slot = `unknown:${unknownNonce}`;
+    sharedData.byVal.init(val, slot);
+    sharedData.bySlot.init(slot, val);
+    return slot;
+  };
+
+  const makeSaver = (kind, tables) => {
+    let nonce = 0;
+    return val => {
+      const slot = `${(nonce += 1)}`;
+      tables.bySlot.init(slot, val);
+      tables.byVal.init(val, slot);
+    };
+  };
+
+  return harden({
+    savePurseActions: makeSaver('purse', myData.purse),
+    savePaymentActions: makeSaver('payment', myData.payment),
+    /**
+     * @param {string} id
+     * @param {Purse} purse
+     */
+    initPurseId: (id, purse) => {
+      myData.purse.bySlot.init(id, purse);
+      myData.purse.byVal.init(purse, id);
+    },
+    purseEntries: myData.purse.bySlot.entries,
+    /**
+     * @param {string} id
+     * @param {unknown} val
+     */
+    initBoardId: (id, val) => {
+      sharedData.bySlot.init(id, val);
+      sharedData.byVal.init(val, id);
+    },
+    /**
+     * @param {string} id
+     * @param {unknown} val
+     */
+    ensureBoardId: (id, val) => {
+      if (sharedData.byVal.has(val)) {
+        assert.equal(sharedData.byVal.get(val), id);
+        return;
+      }
+      sharedData.bySlot.init(id, val);
+      sharedData.byVal.init(val, id);
+    },
+    ...makeMarshal(valToSlot, slotToVal),
+  });
+};
+
+const defaultMakePresence = iface => {
+  const severed = `SEVERED: ${iface.replace(/^Alleged: /, '')}`;
+  const thing = /** @type {any} */ (Far(severed, {}));
+  return thing;
+};
+
+/**
+ * Make context for unserializing wallet or board data.
+ *
+ * @param {(iface: string) => unknown} [makePresence]
+ */
+export const makeImportContext = (makePresence = defaultMakePresence) => {
+  // constraint: none of these keys has another as a prefix.
+  // and none overlaps with board
+  const myData = {
+    purse: {
+      /** @type {MapStore<string, PurseActions>} */
+      bySlot: makeScalarMap(),
+      /** @type {MapStore<PurseActions, string>} */
+      byVal: makeScalarMap(),
+    },
+    payment: {
+      /** @type {MapStore<string, PaymentActions>} */
+      bySlot: makeScalarMap(),
+      /** @type {MapStore<PaymentActions, string>} */
+      byVal: makeScalarMap(),
+    },
+    // TODO: 6 in total, right?
+  };
+  const sharedData = {
+    /** @type {MapStore<string, unknown>} */
+    bySlot: makeScalarMap(),
+    /** @type {MapStore<unknown, string>} */
+    byVal: makeScalarMap(),
+  };
+
+  const kindOf = slot => Object.keys(myData).find(k => slot.startsWith(k));
+
+  const slotToVal = {
+    /**
+     * @param {string} slot
+     * @param {string} iface
+     */
+    fromBoard: (slot, iface) => {
+      if (sharedData.bySlot.has(slot)) {
+        return sharedData.bySlot.get(slot);
+      }
+      const kind = kindOf(slot);
+      if (kind) {
+        assert.fail(X`bad shared slot ${q(slot)}`);
+      }
```

Comment:
I would expect:
```suggestion
      assert(!kind, X`bad shared slot ${q(slot)}`);
```
But does the `.fail` have a different signal?

I'm not clear on how it could get to this state. Is this one of those asserts that should impossible by design and is a safeguard here? If so, the message should say so.

If not, could the error help the developer to solve?